plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+	if timestamp == nil {
+		return p.TimeFunc(), nil
+	}
+
+	if p.TimestampFormat == "" {
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
+
+	metricTime, err := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
+	if err != nil {
+		return p.TimeFunc(), err
+	}
+
+	return metricTime, nil
+}
+
+func (p *Parser) createMetric(native interface{}, schema string) (telegraf.Metric, error) {
+	deep, ok := native.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("cannot cast native interface {} to map[string]interface{}")
+	}
+
+	metricTime, err := p.parseTimestamp(nestedValue(deep[p.Timestamp]))
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
+		} else {
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
+	}
+	fieldList := make([]string, len(p.Fields), (cap(p.Fields)+1)*2)
+	copy(fieldList, p.Fields)
+	if len(fieldList) == 0 { // Get fields from whatever we just unpacked
+		for k := range deep {
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	for _, field := range fieldList {
+		if value, ok := deep[field]; ok {
+			fields[field] = nestedValue(value)
+		} else {
+			p.Log.Infof("field %s value was %v; not added to fields", field, value)
+		}
+	}
+	var schemaObj map[string]interface{}
+
+	err = json.Unmarshal([]byte(schema), &schemaObj)
+	if err != nil {
+		return nil, err
+	}
+	if len(fields) == 0 {
+		return nil, fmt.Errorf("number of fields is 0; unable to create metric")
+	}
+	name := ""
+	if p.Measurement != "" {
+		name = p.Measurement
+	} else {
+		// get Measurement name from schema
+		nsStr, ok := schemaObj["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine namespace from schema %s", schema)
+		}
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + "." + nStr
+	}
+	if name == "" {
+		name = p.MetricName
+	}
+	if name == "" {
+		return nil, fmt.Errorf("could not determine measurement name")
+	}
+	if p.DiscardArrays {
+		// Any non-scalars end up being a nil field
+		return metric.New(name, tags, fields, metricTime), nil
+	}
+	// But if we do it this way, we flatten any compound structures,
+	// including arrays.  Goavro is only going to hand us back
+	// arrays, not maps.
+	// The default (the separator string is empty) is equivalent to
+	// what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	flat, err := flatten.Flatten(fields, "", sep)
+	if err != nil {
+		return nil, err
+	}
+	return metric.New(name, tags, flat, metricTime), nil
```

Comment:
Shouldn't we have two distinct function switched at `Init()` time instead of first fill the fields and then throw them away?

Reply:
We don't necessarily know ahead of time if a field is going to be an array or not (because each record we get is a tuple of schema and data, and the schema is often nothing we have met before).

If you just throw an array field at telegraf/metric it doesn't know what to do with it so you just don't get that field in the output metric.  This *might* be what a user wants, but it's certainly not what my use case requires: I want the arrays flattened into multiple columns.  Since I know that *eventually* I want the column names with an underscore separator, but right now (to maintain compatibility with our streamreactor implementation while we transition from it + InfluxDBv1 to the avro parser + InfluxDBv2) I do not, so I made that a configurable choice.

Either way we're calling metric.New() ; and while I guess we could examine each of our fields and determine whether metric.New() is going to drop that field (because it's an array) and pre-discard it, there doesn't seem to me to be a lot of point in doing so.

We could also just make the decision that if you get an array field of course you want it flattened into multiple columns.  That's what streamreactor does, and it's what I want, and I suspect that it is the overwhelmingly-more-common use case, but I'm not sure we want to remove the flatten-or-discard choice from the user entirely.

Reply:
Understood. Seems like I overlooked the dynamic of this thing. :-) Anyway, please split this function into an "for scalars" and "for arrays" function and call the relevant one. For me it is confusing if first fields are set and later be overwritten without any clear indication of what's going on.

Reply:
Done.